ipa-core/src/protocol/context/dzkp_validator.rs

```diff
@@ -0,0 +1,765 @@
+        prss_left: SegmentEntry<'a>,
+        prss_right: SegmentEntry<'a>,
+        z_right: SegmentEntry<'a>,
+    ) -> Self {
+        // check for consistent length
+        // using debug assert for efficiency
+        // debug assert is sufficient since len only depends on the implementation,
+        // i.e. choice of vectorization dimension and Boolean Array type,
+        // user inputs are not a concern as long as it works for all vectorization options
+        debug_assert_eq!(x_left.len(), x_right.len());
+        debug_assert_eq!(x_left.len(), y_left.len());
+        debug_assert_eq!(x_left.len(), y_right.len());
+        debug_assert_eq!(x_left.len(), prss_left.len());
+        debug_assert_eq!(x_left.len(), prss_right.len());
+        debug_assert_eq!(x_left.len(), z_right.len());
+        // check that length is either multiple of 256 or 256 is multiple of length
+        debug_assert!(x_left.len() % 256 == 0 || 256 % x_left.len() == 0);
+        // asserts passed, create struct
+        Self {
+            x_left,
+            x_right,
+            y_left,
+            y_right,
+            prss_left,
+            prss_right,
+            z_right,
+        }
+    }
+
+    #[must_use]
+    pub fn len(&self) -> usize {
+        self.x_left.len()
+    }
+
+    #[must_use]
+    pub fn is_empty(&self) -> bool {
+        self.x_left.is_empty()
+    }
+}
+
+/// `SegmentEntry` is a simple wrapper to represent one entry of a `Segment`
+/// currently, we only support `BitSlices`
+#[derive(Clone, Debug)]
+pub struct SegmentEntry<'a>(&'a BitSlice<u8, Lsb0>);
+
+impl<'a> SegmentEntry<'a> {
+    #[must_use]
+    pub fn new(entry: &'a BitSlice<u8, Lsb0>) -> Self {
+        SegmentEntry(entry)
+    }
+
+    #[must_use]
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    #[must_use]
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+}
+
+/// `UnverifiedValuesStore` stores a vector of `UnverifiedValues` together with an `offset`.
+/// `offset` is the minimum `RecordId` for the current batch.
+/// `chunk_size` is the amount of records within a single batch. It is used to estimate the vector length
+/// during the allocation.
+/// `segment_size` is an option and set to `Some` once we receive the first segment and thus know the size.
+///
+/// `vec` is a `Vec<Option>` such that we can initialize the whole vector with `None` and access any
+/// location out of order. This is important since we do not now the order in which different records
+/// are added to the `UnverifiedValueStore`.
+/// It is not to keep track whether all relevant records have been added. The verification will fail
+/// when a helper party forgets to include one of the records. We could add a check separate from the proof
+/// such that this does not count as malicious behavior.
+#[derive(Clone, Debug)]
+struct UnverifiedValuesStore {
+    offset: RecordId,
+    chunk_size: usize,
+    segment_size: Option<usize>,
```

Comment:
what is the reason for having variable at runtime size here? Is it possible to simplify this implementation by

* Assuming the size is known in advance
* Allocating early
* Having a counter that tracks the rightmost position of a non-empty segment. This one may be important because you have quite a few methods that perform linear scan and for large stores which is what we want the cost will be significant